Add test which attempts to modify static memory at compile-time

No matter what kind of const eval features we add, we should never allow this. Even if we managed to allow this, I think the current impl would ICE. Even if it does not ICE, it could never actually modify the memory. We should ensure that the following code always errors.

```rust
#![feature(const_raw_ptr_deref)]
#![feature(const_let)]

use std::cell::UnsafeCell;

struct Foo(UnsafeCell<u32>);

unsafe impl Send for Foo {}
unsafe impl Sync for Foo {}

static FOO: Foo = Foo(UnsafeCell::new(42));

static BAR: () = unsafe {
    *FOO.0.get() = 5;
};

fn main() {
    println!("{}", unsafe { *FOO.0.get() });
}
```

Hi @oli-obk ,
When I was running the test in step 3, or compiled the new test with nightly rustc (rust-lang/rust@fc81e3624), I got errors like this.

error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
  --> mod-static-with-const-fn.rs:24:6
   |
24 |     *FOO.0.get() = 5;
   |      ^^^^^^^^^^^

error[E0658]: statements in statics are unstable (see issue #48821)
  --> mod-static-with-const-fn.rs:24:5
   |
24 |     *FOO.0.get() = 5;
   |     ^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(const_let)] to the crate attributes to enable

error: aborting due to 2 previous errors

I think E0015 is the expected error. But I'm not sure about E0658. It seems that the const_let feature is not stabilized yet. (But compiling const_let.rs doesn't return this error.) Could you please help with this problem. Thanks a lot.

--- a/src/test/ui/consts/const-eval/mod-static-with-const-fn.rs
+++ b/src/test/ui/consts/const-eval/mod-static-with-const-fn.rs
@@ -1,0 +1,37 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// New test for #53818: modifying static memory at compile-time is not allowed.
+// The test should never succeed.
+
+#![feature(const_raw_ptr_deref)]
+#![feature(const_let)]
+
+use std::cell::UnsafeCell;
+
+struct Foo(UnsafeCell<u32>);
+
+unsafe impl Send for Foo {}
+unsafe impl Sync for Foo {}
+
+static FOO: Foo = Foo(UnsafeCell::new(42));
+
+static BAR: () = unsafe {
+    *FOO.0.get() = 5;
+    //~^ ERROR calls in statics are limited to constant functions, tuple structs and tuple variants
+
+    // This error is caused by a separate bug that the feature gate error is reported
+    // even though the feature gate "const_let" is active.
+    //~| statements in statics are unstable (see issue #48821)
+};
+
+fn main() {
+    println!("{}", unsafe { *FOO.0.get() });
+}

--- a/src/test/ui/consts/const-eval/mod-static-with-const-fn.stderr
+++ b/src/test/ui/consts/const-eval/mod-static-with-const-fn.stderr
@@ -1,0 +1,18 @@
+error[E0015]: calls in statics are limited to constant functions, tuple structs and tuple variants
+  --> $DIR/mod-static-with-const-fn.rs:27:6
+   |
+LL |     *FOO.0.get() = 5;
+   |      ^^^^^^^^^^^
+
+error[E0658]: statements in statics are unstable (see issue #48821)
+  --> $DIR/mod-static-with-const-fn.rs:27:5
+   |
+LL |     *FOO.0.get() = 5;
+   |     ^^^^^^^^^^^^^^^^
+   |
+   = help: add #![feature(const_let)] to the crate attributes to enable
+
+error: aborting due to 2 previous errors
+
+Some errors occurred: E0015, E0658.
+For more information about an error, try `rustc --explain E0015`.

